### PR TITLE
Proxy handler dev

### PR DESCRIPTION
**Pull request by adisri**
Hey guys! We have completed the ProxyHandler for Assignment 7. Just wanted to let you know that your Config test has been failing for us from the start, and we are not sure why. Apart from that, everything else (including our additions) are passing tests. Thanks for reviewing our pull request!

### CODE REVIEW

File: #Makefile# [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+gcov: test
+	gcov -o $(OBJ_DIR) -r src/*
+
+.PHONY: clean
+clean:
+	$(MAKE) -C src clean
+	$(MAKE) -C test clean
+	$(MAKE) -C lib clean
+	rm -rf $(OBJ_DIR)/*.gcda $(OBJ_DIR)/*.gcno
+z
```
Comment (jobu95) [inline #Makefile#]: Why does this file exist?

Comment (u3amar) [reply #Makefile#]: emacs buffer. Deleting now

File: bin/resources/base_config~ [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  root /opt/static;
+}
+
+path /echo EchoHandler {}
+
+path /status StatusHandler {}
+
+# Default response handler if no handlers match.
+# TODO this doesn't work yet
+default NotFoundHandler {}
```
Comment (jobu95) [inline bin/resources/base_config~]: This file also appears to be extraneous

Comment (u3amar) [reply bin/resources/base_config~]: Will delete. 

Unrelated - need to talk to you real quick about Session::do_write. It looks like it doesn't handle responses more than 64 kb large. Is there any easy fix for this? The images we're loading from websites are all super pixelated.

File: src/handler/proxy_handler.cc [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  }
+  
+  std::unique_ptr<Request> proxy_req = CreateProxyRequestFromClientRequest(request, host_name);
+  
+  try {
+    IssueProxyRequestAndGetResponse(host_name, host_port, *proxy_req, response);
+  } catch (std::exception& e) {
+    std::cerr << "Error sending proxy request: " << e.what() << std::endl;
+    return RequestHandler::Status::GENERIC_ERROR;
+  }
```
Comment (jobu95) [inline src/handler/proxy_handler.cc:64]: In the rest of our codebase we don't really use C++ exceptions, but it's not a big issue. @ahesselgrave we may want to refactor this function in the future.

Comment (jobu95) [reply src/handler/proxy_handler.cc:64]: other team: no need to worry about it

File: src/handler/proxy_handler.cc [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      if (move_loc.find(http) == 0) {
+	move_loc = move_loc.replace(0, https.length(), "");
+      }
+      
+      return move_loc;
+    }
+  } 
+
+  return "";
+}
```
Comment (jobu95) [inline src/handler/proxy_handler.cc]: LGTM

**Comment by jobu95** [comment]
Once merge conflicts are resolved and @ahesselgrave approves, I'm ready to merge.

**Comment by ahesselgrave** [comment]
Looks solid to me. :shipit:

**Review by ahesselgrave — APPROVED** [review]

**Comment by u3amar** [comment]
We're making a small fix to an integration test, so when that commit comes through we should be good to merge the PR. Should be in less than half an hour or so. Thanks guys!

**Comment by u3amar** [comment]
Hey guys - this is ready to merge into master now. Thanks.

**Comment by jobu95** [comment]
All done?

**Comment by u3amar** [comment]
Almost - sorry for the delay. We just found out we had to add a bunch of unit tests/integration tests for the submission. Will be done in an hour or so.

**Comment by u3amar** [comment]
Ready to merge! Thanks guys.

**Comment by u3amar** [comment]
Hey guys, just a reminder to merge as soon as you get the chance. We don't want to get hit with the 4% penalty =p

**Pull request merged**